Sources/GRPC/ClientConnection.swift

```diff
@@ -53,139 +53,239 @@ import NIOTLS
+          self.connectivity.state = .ready
         }
+        self.replaceChannelAndMultiplexerOnClose(channel: newChannel)
       }
-
-    return bootstrap
+    }
   }
 
-  /// Verifies that a TLS handshake was successful by using the `TLSVerificationHandler`.
-  ///
-  /// - Parameter channel: The channel to verify successful TLS setup on.
-  public class func verifyTLS(channel: Channel) -> EventLoopFuture<Void> {
-    return channel.pipeline.handler(type: TLSVerificationHandler.self).flatMap {
-      $0.verification
+  /// Replaces the `channel` and `multiplexer` on this instance and registers a callback on the
+  /// `closeFuture` of the newly created `channel` to call this function.
+  private func replaceChannelAndMultiplexer() {
+    let newChannel = ClientConnection.makeChannel(
+      configuration: self.configuration,
+      connectivityMonitor: self.connectivity
+    )
+
+    newChannel.always { result in
+      if case .failure = result {
+        self.connectivity.state = .shutdown
+      }
+    }.flatMap {
+      $0.closeFuture
+    }.whenComplete { _ in
+      // `.shutdown` is terminal: don't attempt a reconnection.
+      if self.connectivity.state != .shutdown {
+        self.replaceChannelAndMultiplexer()
+      }
     }
-  }
 
-  /// Makes a `ClientConnection` from the given channel and configuration.
-  ///
-  /// - Parameter channel: The channel to use for the connection.
-  /// - Parameter configuration: The configuration used to create the channel.
-  public class func makeClientConnection(
-    channel: Channel,
-    configuration: Configuration
-  ) -> EventLoopFuture<ClientConnection> {
-    return channel.pipeline.handler(type: HTTP2StreamMultiplexer.self).map { multiplexer in
-      ClientConnection(channel: channel, multiplexer: multiplexer, configuration: configuration)
+    self.channel = newChannel
+    self.multiplexer = newChannel.flatMap {
+      $0.pipeline.handler(type: HTTP2StreamMultiplexer.self)
+    }
+
+    // Change the state if the connection was successful.
+    newChannel.whenSuccess { _ in
+      self.connectivity.state = .ready
     }
   }
 
-  /// Starts a client connection using the given configuration.
-  ///
-  /// This involves: creating a `ClientBootstrap`, connecting to a target, verifying that the TLS
-  /// handshake was successful (if TLS was configured) and creating the `ClientConnection`.
-  /// See the individual functions for more information:
-  ///  - `makeBootstrap(configuration:)`,
-  ///  - `verifyTLS(channel:)`, and
-  ///  - `makeClientConnection(channel:configuration:)`.
-  ///
-  /// - Parameter configuration: The configuration to start the connection with.
-  public class func start(_ configuration: Configuration) -> EventLoopFuture<ClientConnection> {
-    return start(configuration, backoffIterator: configuration.connectionBackoff?.makeIterator())
+  /// The `EventLoop` this connection is using. Note that this _may_ change over time.
```

Comment:
Not a fan of allowing this to change; could we re-use the previous channel's event loop instead?

Reply:
Do you mean re-use it for the new channel? Or store it and always return it here? Because the former isn't possible.

Reply:
I was hoping for the former; why would that be a problem?

(My concern is that Vapor does a lot of thread-local stuff and generally expects all "sub-requests" to run on the same event loop the original request is handled on. Open to keep this as-is for now, though, given that one could simply pass in the request's `EventLoop` as the client's `EventLoopGroup` argument.)

Reply:
I take that back -- I didn't realise `EventLoop` conformed to `EventLoopGroup`!